staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapiv3/controller.go

```diff
@@ -0,0 +1,295 @@
+
+	if !cache.WaitForCacheSync(stopCh, c.crdsSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	crds, err := c.crdLister.List(labels.Everything())
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to initially list all CRDs: %v", err))
+		return
+	}
+	for _, crd := range crds {
+		if !apiextensionshelpers.IsCRDConditionTrue(crd, apiextensionsv1.Established) {
+			continue
+		}
+		for _, v := range crd.Spec.Versions {
+			if !v.Served {
+				continue
+			}
+			c.buildV3Spec(crd, crd.Name, v.Name)
+		}
+	}
+
+	// only start one worker thread since its a slow moving API
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	// log slow aggregations
+	start := time.Now()
+	defer func() {
+		elapsed := time.Since(start)
+		if elapsed > time.Second {
+			klog.Warningf("slow openapi aggregation of %q: %s", key.(string), elapsed)
+		}
+	}()
+
+	err := c.syncFn(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	c.queue.AddRateLimited(key)
+	return true
+}
+
+func (c *Controller) sync(name string) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	crd, err := c.crdLister.Get(name)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	if errors.IsNotFound(err) || !apiextensionshelpers.IsCRDConditionTrue(crd, apiextensionsv1.Established) {
+		c.deleteCRD(name)
+		return nil
+	}
+
+	for _, v := range crd.Spec.Versions {
+		if !v.Served {
+			continue
+		}
+		c.buildV3Spec(crd, name, v.Name)
+	}
+
+	return nil
+}
+
+func (c *Controller) deleteCRD(name string) {
+	for gvString, crdList := range c.crdSpecsByGV {
+		needOpenAPIUpdate := false
+		for crdName, crd := range crdList {
+			if crd.name == name {
+				needOpenAPIUpdate = true
+				delete(crdList, crdName)
+			}
+		}
+		if needOpenAPIUpdate {
+			if len(crdList) == 0 {
+				delete(c.crdSpecsByGV, gvString)
+			}
+			c.updateGroupVersion(gvString)
+		}
+	}
+}
+
+func (c *Controller) updateGroupVersion(gvString string) error {
+	if _, ok := c.crdSpecsByGV[gvString]; !ok {
+		c.openAPIV3Service.DeleteGroupVersion(gvString)
+		return nil
+	}
+
+	var specs []*spec3.OpenAPI
+	for _, crd := range c.crdSpecsByGV[gvString] {
+		specs = append(specs, crd.spec)
+	}
+
+	mergedSpec, err := builder.MergeSpecsV3(specs...)
+	if err != nil {
+		return fmt.Errorf("failed to merge specs: %v", err)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if c.openAPIV3Service != nil {
+		c.openAPIV3Service.UpdateGroupVersion(gvString, mergedSpec)
+	}
+	return nil
+}
+
+func (c *Controller) updateV3CRD(crd *apiextensionsv1.CustomResourceDefinition, name, versionName string, v3 *spec3.OpenAPI) error {
```

Comment:
Can we rename that "v3" variable here? maybe `newSpec` or something?

Reply:
unresolved